src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenSubscriptions.cs

```diff
@@ -25,13 +35,42 @@ protected internal override void Setup(FeatureConfigurationContext context)
                 throw new Exception("The selected persistence doesn't have support for subscription storage. Select another persistence or disable the message-driven subscriptions feature using endpointConfiguration.DisableFeature<MessageDrivenSubscriptions>()");
             }
 
-            context.Pipeline.Register<SubscriptionReceiverBehavior.Registration>();
-            var authorizer = context.Settings.GetSubscriptionAuthorizer();
-            if (authorizer == null)
+            var transportInfrastructure = context.Settings.Get<TransportInfrastructure>();
+            var canReceive = !context.Settings.GetOrDefault<bool>("Endpoint.SendOnly");
+            var conventions = context.Settings.Get<Conventions>();
+            var enforceBestPractices = context.Settings.Get<bool>(BestPracticesFeature.EnforceBestPracticesSettingsKey);
+
+            var distributionPolicy = context.Settings.Get<DistributionPolicy>();
+            var endpointInstances = context.Settings.Get<EndpointInstances>();
+            var publishers = context.Settings.Get<Publishers>();
+            var configuredPublishers = context.Settings.Get<ConfiguredPublishers>();
+
+            configuredPublishers.Apply(publishers, conventions, enforceBestPractices);
+
+            context.Pipeline.Register(b =>
             {
-                authorizer = _ => true;
+                var unicastPublishRouter = new UnicastPublishRouter(b.Build<MessageMetadataRegistry>(), b.Build<ISubscriptionStorage>());
+                return new UnicastPublishRouterConnector(unicastPublishRouter, distributionPolicy);
+            }, "Determines how the published messages should be routed");
+
+            if (canReceive)
+            {
+                var distributorAddress = context.Settings.GetOrDefault<string>("LegacyDistributor.Address");
+
+                var subscriberAddress = distributorAddress ?? context.Settings.LocalAddress();
+                var subscriptionRouter = new SubscriptionRouter(publishers, endpointInstances, i => transportInfrastructure.ToTransportAddress(LogicalAddress.CreateRemoteAddress(i)));
+
+                context.Pipeline.Register(b => new MessageDrivenSubscribeTerminator(subscriptionRouter, subscriberAddress, context.Settings.EndpointName(), b.Build<IDispatchMessages>()), "Sends subscription requests when message driven subscriptions is in use");
```

Comment:
It would be also good to move `UnicastPublishRouterConnector` also here as it depends on the subscription store which is message-driven pub/sub concept.